chess-AI/ai_io.py

```diff
@@ -2,63 +2,136 @@
+    will be loaded into the path.
+    """
+    # See if dataset should be loaded from Figshare
+    if ds_saving['figshare_load']:
+        # Get the base path and file name
+        file_name = ds_saving['file_name']
+        # ensure file exists in figshare
+        assert figshare_api.get_figshare_article(ds_saving['data_dir'], file_name), \
+               "File not found in figshare"
+        msg = f"{file_name} downloaded from figshare"
+    else:
+        # Load the dataset at the provided path. If using streamlit,
+        # show a confirmation message.
+        assert os.path.exists(ds_saving['data_dir']), "Dataset not found at path provided." \
+               "You must provide a path to an existing dataset"
+        full_path = ds_saving['data_dir'] + ds_saving['file_name']
+        msg = f"dataset from {full_path} retrieved"
 
-    msg = "Dataset retrieved."
     if show_dash:
         Dashboard.info_message("success", msg)
     else:
         print(msg)
+    # load the dataset into the path specified
+    return torch.load(ds_saving['data_dir'] + ds_saving['file_name'])
 
-    return torch.load(dataset_path)
 
+def save_dataset(dataset, ds_saving):
+    """Save a dataset to Figshare or Locally
 
-def save_model(nnet, save_path, num_saved_models, overwrite):
-    """Save given model parameters to external file
+    The save_dir refers to the directory where the file
+    will be saved. For saving to Figshare, the file still needs to
+    be saved locally first.
     """
-    if save_path is not None:
-        torch.save(nnet.state_dict(), save_path)
-    else:
-        # Iterate through the number of models saved
-        for i in range(num_saved_models):
-            if not os.path.isfile(f'./models/models-{i + 1}.pt'):
-                if overwrite and i != 0:
-                    torch.save(nnet.state_dict(), f'./models/models-{i}.pt')
-                    break
-                torch.save(nnet.state_dict(), f'./models/models-{i + 1}.pt')
-                break
-            if i == num_saved_models - 1:
-                torch.save(nnet.state_dict(), f'./models/models-{num_saved_models}.pt')
-
-
-def load_model(nnet, model_path, num_saved_models):
-    """Load model parameters into given network from external file
+    # make sure a path was specified
+    date_string = create_date_string()
+    # get the full file path to save
+    full_path = ds_saving['data_dir'] + "dataset-" +  date_string + ".pt"
+    # save the dataset
+    torch.save(dataset, full_path)
+
+    if ds_saving['figshare_save']:
+        # Save dataset to figshare
+        title = "dataset-" + date_string
+        desc = "This description is a placeholder"
+        keys = ["Dataset", "Chess"]
+        # 179 is category [Artificial Intelligence and Image Processing]
+        categories = [179]
+        figshare_api.upload(title, desc, keys, categories, full_path)
+
+
+def save_model(nnet, m_saving, checkpointing, file_name=None):
```

Comment:
Can you update the docstring to describe what `m_saving` is? I don't know from looking at this what all the keys are.